text/0000-debug-noop-in-release-builds.md

```diff
@@ -0,0 +1,30 @@
+- Feature Name: debug-noop-in-release
+- Start Date: 2014-03-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+The `{:?}` format specifier should not print values in release builds.
+
+# Motivation
+
+Many commonly used methods, such as `unwrap`, use `{:?}` to print out representations of values if their conditions are not met. This is useful for debugging but is not so useful in release builds, and it can bloat up the generated binaries. Worse, it adds opaque calls (as `Debug` is frequently a lot of code and rarely gets inlined), which can mess up the inlining heuristics and cause cascades of missed optimization opportunities.
```

Comment:
It sounds like the solution here is to use a debug_assert or something, rather than changing the behaviour of `{:?}`. If this is done thoroughly then the `Debug` impls should be dead code and removed from binaries. (It should be fairly simple to add tool support for this sort of thing).